# iris dataset - Decision Tree implemantation and Bagging

Use Decision Tree to create model for Iris Dataset:
- DecisionTreeClassifier


If you label column is categorical use DecisionTreeClassifier 

If you label column is continuos numberical use DecisionTreeRegressor

In [1]:
import numpy as np
import pandas as pd

In [2]:
data = pd.read_csv('https://gist.githubusercontent.com/curran/a08a1080b88344b0c8a7/raw/639388c2cbc2120a14dcf466e85730eb8be498bb/iris.csv')

In [3]:
data.head()

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   sepal_length  150 non-null    float64
 1   sepal_width   150 non-null    float64
 2   petal_length  150 non-null    float64
 3   petal_width   150 non-null    float64
 4   species       150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB


In [5]:
features =data.iloc[:, :-1].values
label = data.iloc[:, -1].values

In [6]:
#Train Test Split
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(features,
                                                    label,
                                                    test_size=0.2,
                                                    random_state=23)

In [7]:
# DecisionTreeClassifier ==> classification usecase
# DecisionTreeRegressor ==> Regression usecase

from sklearn.tree import DecisionTreeClassifier
modelTree = DecisionTreeClassifier()
modelTree.fit(X_train, y_train)

DecisionTreeClassifier()

In [8]:
print(modelTree.score(X_train, y_train))
print(modelTree.score(X_test, y_test))

1.0
0.9666666666666667


train_score is 1.0

test_score is 0.9666666666666667

Beacuse train_score > test_score is an overfitted model and we need to fix it.

First: let play with max_depth and see we can fix it or not.

In [10]:
# max_depth=6
from sklearn.tree import DecisionTreeClassifier
modelTree = DecisionTreeClassifier(max_depth=6)
modelTree.fit(X_train, y_train)

print(modelTree.score(X_train, y_train))
print(modelTree.score(X_test, y_test))

1.0
0.9666666666666667


In [11]:
#max_depth=5
modelTree = DecisionTreeClassifier(max_depth=5)
modelTree.fit(X_train, y_train)

print(modelTree.score(X_train, y_train))
print(modelTree.score(X_test, y_test))

0.9916666666666667
0.9666666666666667


In [12]:
#max_depth=4
modelTree = DecisionTreeClassifier(max_depth=4)
modelTree.fit(X_train, y_train)

print(modelTree.score(X_train, y_train))
print(modelTree.score(X_test, y_test))

0.9833333333333333
0.9666666666666667


If we use for loop and play with modelTree = DecisionTreeClassifier(max_depth=i) , we might get an ideal model

In [14]:
SL = 0.1
CL = 1 - SL

for i in range(1,100):
    X_train, X_test, y_train, y_test = train_test_split(features,
                                                    label,
                                                    test_size=0.2,
                                                    random_state=i)
    modelTree = DecisionTreeClassifier(max_depth=4)
    modelTree.fit(X_train, y_train)

    train_score = modelTree.score(X_train,y_train)
    test_score = modelTree.score(X_test,y_test)

    if test_score > train_score and test_score > CL:
        print("Test Score {}, Train Score {}, Random_state {}".format(test_score,train_score,i))
    

Test Score 1.0, Train Score 0.9916666666666667, Random_state 3
Test Score 1.0, Train Score 0.9916666666666667, Random_state 9
Test Score 1.0, Train Score 0.9916666666666667, Random_state 19
Test Score 1.0, Train Score 0.9916666666666667, Random_state 35
Test Score 1.0, Train Score 0.9833333333333333, Random_state 37
Test Score 1.0, Train Score 0.9916666666666667, Random_state 38
Test Score 1.0, Train Score 0.9916666666666667, Random_state 40
Test Score 1.0, Train Score 0.975, Random_state 42
Test Score 1.0, Train Score 0.9916666666666667, Random_state 57
Test Score 1.0, Train Score 0.9916666666666667, Random_state 70
Test Score 1.0, Train Score 0.9916666666666667, Random_state 72
Test Score 1.0, Train Score 0.975, Random_state 74
Test Score 1.0, Train Score 0.9916666666666667, Random_state 75
Test Score 1.0, Train Score 0.9916666666666667, Random_state 81
Test Score 1.0, Train Score 0.9916666666666667, Random_state 84
Test Score 1.0, Train Score 0.9916666666666667, Random_state 91
Test

Test Score 1.0, Train Score 0.975, Random_state 42

Random_state 42 is getting us something. 

In [15]:
#Train Test Split
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(features,
                                                    label,
                                                    test_size=0.2,
                                                    random_state=42)

In [16]:
modelTree = DecisionTreeClassifier(max_depth=5)
modelTree.fit(X_train, y_train)

print(modelTree.score(X_train, y_train))
print(modelTree.score(X_test, y_test))

0.9916666666666667
1.0


train_score is 0.9916666666666667

test_score is 1.0

Beacuse train_score < test_score we have a generalized model